src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -205,13 +261,62 @@ protected AttributeAccessor getErrorMessageAttributes(Message<?> message) {
 		}
 	}
 
+	private void registerSeekCallbackIfNecessary(ConsumerSeekAware.ConsumerSeekCallback callback) {
+		if (this.setAdditionalHeaders) {
+			if (this.messageListenerContainer instanceof ConcurrentMessageListenerContainer) {
+
+				seekCallbackHolder.set(callback);
+			}
+			else {
+				this.seekCallback = callback;
+			}
+		}
+	}
```

Comment:
Not sure we need this extra complexity - why not always use the ThreadLocal?

Reply:
Well, it's not end-user API, so I don't see reason to worry about complexity in our code since with the single-threaded container we will avoid `ThreadLocal` overhead altogether.